### PR TITLE
Allowing generic item type in new experimental SelectControl

### LINKED ISSUE #34399: [Components] Allow generic item type in new experimental SelectControl
A search component has been created as part of https://github.com/woocommerce/woocommerce/pull/34159

Currently it only takes the `ItemType` as format, which is of type: `{ value: string; label: string }`
We should update the search component to accept any object type and handle the label and value/key properties by optional callback functions.
There is already one for the label - `itemToString` and we should create one for the `key` maybe `getItemKey` ( and maybe convert `itemToString` to `getItemLabel` to keep consistency).

To have this play well with TypeScript we will have to allow users to pass a generic to the component, so one could use it like this: `<SelectControl<RandomType> .... />`
See the `<Form />` component as an example: https://github.com/woocommerce/woocommerce/blob/trunk/packages/js/components/src/form/form.tsx#L80 (this passes the generic `Values` prop also into the Property type. I don't think we need to make use of `extends` though, and should just default to `ItemType`, something like: `function SelectControl< Item = ItemType>`

**Acceptance Criteria**

* [ ] SelectControl should allow for a custom Typescript type
* [ ] SelectControl should make use of the label and key helper functions in any place it uses one of these.

### PR DESCRIPTION
**Pull request by joelclimbsthings**
-   [x] Have you followed the [WooCommerce Contributing guideline](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md)?
-   [x] Does your code follow the [WordPress' coding standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/)?
-   [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change?

### Changes proposed in this Pull Request:

Updates to the new experimental SelectControl to support passing a custom type to describe the items, as well as including `getItemValue()` and `getItemLabel()` functions as props to parse those items. 

Note: This PR was a bit adventurous for me as I wasn't entirely familiar with TS generic types, but it was a fun learning experience. The only issue I haven't been able to resolve is a couple [TS errors within the utils for the default `item` and `value` getters](https://github.com/woocommerce/woocommerce/pull/34547/files#r960866165). 

Closes #34399 .

### How to test the changes in this Pull Request:

1. Checkout branch.
2. Fire up storybook with `npm run storybook`
3. Observe and interact with the new `Experimental > SelectControl > Custom Item Type` story.
4. Optionally including the Select control with a custom type (can copy from story) live on any react screen to interact with it there.

### Other information:

-   [x] Have you added an explanation of what your changes do and why you'd like us to include them?
-   [ ] Have you written new tests for your changes, as applicable?
-   [x] Have you successfully run tests with your changes locally?
-   [x] Have you created a changelog file for each project being changed, ie `pnpm changelog add --filter=<project>`?

### FOR PR REVIEWER ONLY:

-   [ ] I have reviewed that everything is sanitized/escaped appropriately for any SQL or XSS injection possibilities. I made sure Linting is not ignored or disabled.

### CODE REVIEW

File: packages/js/components/src/experimental-select-control/utils.ts
Comment (joelclimbsthings) [inline packages/js/components/src/experimental-select-control/utils.ts]: Referencing `item.label` and `item.value` below are causing TS errors, since TS isn't aware that they're only intended to be used along with the default item type. Is there any way to remedy this in a TS supported way?

Comment (louwie17) [reply packages/js/components/src/experimental-select-control/utils.ts]: Yeah, you can make us of the `is` type guard -> https://www.typescriptlang.org/docs/handbook/advanced-types.html#user-defined-type-guards
The easiest way to do this is by creating a `isDefaultItemType` function, like so:
```
function isDefaultItemType(item: ItemType | null): item is ItemType {
  return (item as ItemType).label !== undefined;
}
```
You could include the check for `value` here as well if you wanted.
Then in the above code you can do:
```
if ( isDefaultItemType(item) ) {
  return item.label;
} else {
return '';
}
```
TypeScript should not complain about `item.label` within the if statement now.

File: packages/js/components/src/experimental-select-control/types.ts
Comment (louwie17) [inline packages/js/components/src/experimental-select-control/types.ts]: Could we also allow the return of a `number` -> `string | number` for the value type? This is not a must if it will require a bunch of checks and conversions everywhere else, but if we can do it somewhat seamless then that would be really nice.

**Review by louwie17 — COMMENTED** [review]
Really nice work on this @joelclimbsthings, this is testing really well, I merged it into my category field branch locally to try the changes out and things tested well out of the box.

I only have one suggestion for the `getItemValueType` return value to allow for `number` as well, as a lot of `id`'s are numbers.
And I left a response to your TS question here: https://github.com/woocommerce/woocommerce/pull/34547/files#r960866165

**Comment by joelclimbsthings** [comment]
Thanks for the help @louwie17 ! I believe I remedied the above, so let me know what you think. 👂🏻

**Review by louwie17 — DISMISSED** [review]
Thanks for the updates @joelclimbsthings, this tested well and the code looks great, very nice work!!

**Comment by louwie17** [comment]
@joelclimbsthings it does look like your changelog Github action failed with the same error that Josh ran into ( p1663259087459849-slack-C03CPM3UXDJ ), I think it just required a rebase to fix. Let me know if you need a re-approval.

**Review by louwie17 — APPROVED** [review]
Re-approving after rebase, LGTM 🚀

**Pull request merged**
